# Comparing `tmp/pvactools-3.1.2-py3-none-any.whl.zip` & `tmp/pvactools-3.1.3-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,91 +1,91 @@
-Zip file size: 9980384 bytes, number of entries: 785
--rw-r--r--  2.0 unx      445 b- defN 23-Apr-05 14:24 pvactools/lib/__init__.py
--rw-r--r--  2.0 unx    25560 b- defN 23-Apr-05 14:24 pvactools/lib/aggregate_all_epitopes.py
--rw-r--r--  2.0 unx     3099 b- defN 23-Apr-05 14:24 pvactools/lib/allele_specific_binding_filter.py
+Zip file size: 9980395 bytes, number of entries: 785
+-rw-r--r--  2.0 unx      445 b- defN 23-Jun-06 18:07 pvactools/lib/__init__.py
+-rw-r--r--  2.0 unx    25560 b- defN 23-Jun-06 18:07 pvactools/lib/aggregate_all_epitopes.py
+-rw-r--r--  2.0 unx     3099 b- defN 23-Jun-06 18:07 pvactools/lib/allele_specific_binding_filter.py
 -rw-r--r--  2.0 unx      792 b- defN 22-May-13 15:22 pvactools/lib/allele_specific_cutoffs.py
--rw-r--r--  2.0 unx     5537 b- defN 23-Apr-05 14:24 pvactools/lib/binding_filter.py
--rw-r--r--  2.0 unx     3113 b- defN 23-Mar-07 15:29 pvactools/lib/calculate_manufacturability.py
--rw-r--r--  2.0 unx    17927 b- defN 23-Apr-05 14:24 pvactools/lib/calculate_reference_proteome_similarity.py
--rw-r--r--  2.0 unx     2326 b- defN 23-Apr-05 14:24 pvactools/lib/call_iedb.py
--rw-r--r--  2.0 unx     2981 b- defN 23-Apr-05 14:24 pvactools/lib/call_mhcnuggets.py
--rw-r--r--  2.0 unx     2256 b- defN 23-Mar-07 15:29 pvactools/lib/combine_parsed_outputs.py
+-rw-r--r--  2.0 unx     5537 b- defN 23-Jun-06 18:07 pvactools/lib/binding_filter.py
+-rw-r--r--  2.0 unx     3113 b- defN 23-May-25 14:27 pvactools/lib/calculate_manufacturability.py
+-rw-r--r--  2.0 unx    18080 b- defN 23-Jun-06 18:07 pvactools/lib/calculate_reference_proteome_similarity.py
+-rw-r--r--  2.0 unx     2326 b- defN 23-Jun-06 18:07 pvactools/lib/call_iedb.py
+-rw-r--r--  2.0 unx     2981 b- defN 23-Jun-06 18:07 pvactools/lib/call_mhcnuggets.py
+-rw-r--r--  2.0 unx     2256 b- defN 23-May-25 14:27 pvactools/lib/combine_parsed_outputs.py
 -rw-r--r--  2.0 unx     1523 b- defN 22-Nov-04 17:48 pvactools/lib/csq_parser.py
 -rw-r--r--  2.0 unx      955 b- defN 22-May-13 15:22 pvactools/lib/download_example_data.py
 -rw-r--r--  2.0 unx    22842 b- defN 23-Mar-06 21:41 pvactools/lib/fasta_generator.py
--rw-r--r--  2.0 unx     1337 b- defN 23-Apr-05 14:24 pvactools/lib/filter.py
--rw-r--r--  2.0 unx    28534 b- defN 23-Apr-05 14:24 pvactools/lib/input_file_converter.py
--rw-r--r--  2.0 unx    11212 b- defN 23-Apr-05 14:24 pvactools/lib/net_chop.py
--rw-r--r--  2.0 unx    12096 b- defN 23-Apr-05 14:24 pvactools/lib/netmhc_stab.py
+-rw-r--r--  2.0 unx     1337 b- defN 23-Jun-06 18:07 pvactools/lib/filter.py
+-rw-r--r--  2.0 unx    28534 b- defN 23-Jun-06 18:07 pvactools/lib/input_file_converter.py
+-rw-r--r--  2.0 unx    11212 b- defN 23-Jun-06 18:07 pvactools/lib/net_chop.py
+-rw-r--r--  2.0 unx    12096 b- defN 23-Jun-06 18:07 pvactools/lib/netmhc_stab.py
 -rw-r--r--  2.0 unx     3374 b- defN 22-May-13 15:22 pvactools/lib/optimal_peptide.py
--rw-r--r--  2.0 unx    41280 b- defN 23-Apr-05 14:26 pvactools/lib/output_parser.py
--rw-r--r--  2.0 unx    38074 b- defN 23-Apr-05 14:24 pvactools/lib/pipeline.py
--rw-r--r--  2.0 unx     7861 b- defN 23-Apr-05 14:24 pvactools/lib/post_processor.py
--rw-r--r--  2.0 unx    19965 b- defN 23-Apr-05 14:24 pvactools/lib/prediction_class.py
+-rw-r--r--  2.0 unx    41280 b- defN 23-Jun-06 18:07 pvactools/lib/output_parser.py
+-rw-r--r--  2.0 unx    38074 b- defN 23-Jun-06 18:07 pvactools/lib/pipeline.py
+-rw-r--r--  2.0 unx     7861 b- defN 23-Jun-06 18:07 pvactools/lib/post_processor.py
+-rw-r--r--  2.0 unx    19965 b- defN 23-Jun-06 18:07 pvactools/lib/prediction_class.py
 -rw-r--r--  2.0 unx     7569 b- defN 23-Mar-08 18:39 pvactools/lib/proximal_variant.py
 -rw-r--r--  2.0 unx     5818 b- defN 22-May-17 15:15 pvactools/lib/pvacvector_input_fasta_generator.py
--rw-r--r--  2.0 unx    14079 b- defN 23-Apr-05 14:24 pvactools/lib/run_argument_parser.py
--rw-r--r--  2.0 unx     5498 b- defN 23-Apr-05 14:24 pvactools/lib/run_utils.py
--rw-r--r--  2.0 unx     2142 b- defN 23-Apr-05 14:24 pvactools/lib/sort.py
--rw-r--r--  2.0 unx     6962 b- defN 23-Apr-05 14:24 pvactools/lib/top_score_filter.py
+-rw-r--r--  2.0 unx    14079 b- defN 23-Jun-06 18:07 pvactools/lib/run_argument_parser.py
+-rw-r--r--  2.0 unx     5498 b- defN 23-Jun-06 18:07 pvactools/lib/run_utils.py
+-rw-r--r--  2.0 unx     2142 b- defN 23-Jun-06 18:07 pvactools/lib/sort.py
+-rw-r--r--  2.0 unx     6962 b- defN 23-Jun-06 18:07 pvactools/lib/top_score_filter.py
 -rw-r--r--  2.0 unx     1421 b- defN 22-May-13 15:22 pvactools/lib/valid_alleles.py
 -rw-r--r--  2.0 unx     9404 b- defN 22-May-17 15:15 pvactools/lib/vector_visualization.py
 -rw-r--r--  2.0 unx       52 b- defN 22-May-13 15:22 pvactools/tools/__init__.py
 -rw-r--r--  2.0 unx     1094 b- defN 22-May-13 15:22 pvactools/tools/download_cwls.py
 -rw-r--r--  2.0 unx     1082 b- defN 22-May-13 15:22 pvactools/tools/main.py
--rw-r--r--  2.0 unx      285 b- defN 23-Apr-05 14:24 pvactools/tools/pvacbind/__init__.py
+-rw-r--r--  2.0 unx      285 b- defN 23-Jun-06 18:07 pvactools/tools/pvacbind/__init__.py
 -rw-r--r--  2.0 unx      377 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/allele_specific_cutoffs.py
 -rw-r--r--  2.0 unx      495 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/binding_filter.py
--rw-r--r--  2.0 unx      629 b- defN 23-Apr-05 14:24 pvactools/tools/pvacbind/calculate_reference_proteome_similarity.py
+-rw-r--r--  2.0 unx      629 b- defN 23-Jun-06 18:07 pvactools/tools/pvacbind/calculate_reference_proteome_similarity.py
 -rw-r--r--  2.0 unx      376 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/download_example_data.py
--rw-r--r--  2.0 unx     1000 b- defN 23-Apr-05 14:24 pvactools/tools/pvacbind/generate_aggregated_report.py
--rw-r--r--  2.0 unx     3112 b- defN 23-Apr-05 14:24 pvactools/tools/pvacbind/main.py
+-rw-r--r--  2.0 unx     1000 b- defN 23-Jun-06 18:07 pvactools/tools/pvacbind/generate_aggregated_report.py
+-rw-r--r--  2.0 unx     3112 b- defN 23-Jun-06 18:07 pvactools/tools/pvacbind/main.py
 -rw-r--r--  2.0 unx      379 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/net_chop.py
 -rw-r--r--  2.0 unx      394 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/netmhc_stab.py
--rw-r--r--  2.0 unx     7219 b- defN 23-Apr-05 14:24 pvactools/tools/pvacbind/run.py
--rw-r--r--  2.0 unx      384 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/top_score_filter.py
+-rw-r--r--  2.0 unx     7219 b- defN 23-Jun-06 18:07 pvactools/tools/pvacbind/run.py
+-rw-r--r--  2.0 unx      384 b- defN 23-Jun-06 18:07 pvactools/tools/pvacbind/top_score_filter.py
 -rw-r--r--  2.0 unx      358 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/valid_alleles.py
 -rw-r--r--  2.0 unx     3421 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/example_data/input.fasta
 -rw-r--r--  2.0 unx     3882 b- defN 22-May-17 15:14 pvactools/tools/pvacbind/example_data/MHC_Class_I/Test.all_epitopes.aggregated.tsv
 -rw-r--r--  2.0 unx  4191604 b- defN 22-May-17 15:14 pvactools/tools/pvacbind/example_data/MHC_Class_I/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx     5489 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/example_data/MHC_Class_I/Test.filtered.tsv
 -rw-r--r--  2.0 unx     4026 b- defN 22-May-17 15:14 pvactools/tools/pvacbind/example_data/MHC_Class_II/Test.all_epitopes.aggregated.tsv
 -rw-r--r--  2.0 unx   464398 b- defN 22-May-17 15:14 pvactools/tools/pvacbind/example_data/MHC_Class_II/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx     3859 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/example_data/MHC_Class_II/Test.filtered.tsv
 -rw-r--r--  2.0 unx     4017 b- defN 22-May-17 15:14 pvactools/tools/pvacbind/example_data/combined/Test.all_epitopes.aggregated.tsv
 -rw-r--r--  2.0 unx  5052066 b- defN 22-May-17 15:14 pvactools/tools/pvacbind/example_data/combined/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx     6987 b- defN 22-May-13 15:22 pvactools/tools/pvacbind/example_data/combined/Test.filtered.tsv
--rw-r--r--  2.0 unx      315 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/__init__.py
+-rw-r--r--  2.0 unx      315 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/__init__.py
 -rw-r--r--  2.0 unx      377 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/allele_specific_cutoffs.py
 -rw-r--r--  2.0 unx      495 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/binding_filter.py
--rw-r--r--  2.0 unx      629 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/calculate_reference_proteome_similarity.py
+-rw-r--r--  2.0 unx      629 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/calculate_reference_proteome_similarity.py
 -rw-r--r--  2.0 unx      375 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/download_example_data.py
--rw-r--r--  2.0 unx     1000 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/generate_aggregated_report.py
--rw-r--r--  2.0 unx     5051 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/generate_protein_fasta.py
--rw-r--r--  2.0 unx     3420 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/main.py
+-rw-r--r--  2.0 unx     1000 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/generate_aggregated_report.py
+-rw-r--r--  2.0 unx     5051 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/generate_protein_fasta.py
+-rw-r--r--  2.0 unx     3420 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/main.py
 -rw-r--r--  2.0 unx      379 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/net_chop.py
 -rw-r--r--  2.0 unx      394 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/netmhc_stab.py
--rw-r--r--  2.0 unx    13148 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/run.py
--rw-r--r--  2.0 unx      384 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/top_score_filter.py
+-rw-r--r--  2.0 unx    13148 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/run.py
+-rw-r--r--  2.0 unx      384 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/top_score_filter.py
 -rw-r--r--  2.0 unx      358 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/valid_alleles.py
 -rw-r--r--  2.0 unx      761 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/example_data/fusions.bedpe.annot
 -rw-r--r--  2.0 unx     5516 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_I/Test.all_epitopes.aggregated.tsv
 -rw-r--r--  2.0 unx  1172450 b- defN 22-May-20 15:09 pvactools/tools/pvacfuse/example_data/MHC_Class_I/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx     3890 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_I/Test.fasta
 -rw-r--r--  2.0 unx     4981 b- defN 22-May-20 15:09 pvactools/tools/pvacfuse/example_data/MHC_Class_I/Test.filtered.tsv
--rw-r--r--  2.0 unx   338645 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/example_data/MHC_Class_I/10/Test.all_epitopes.final.tsv
--rw-r--r--  2.0 unx   234256 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/example_data/MHC_Class_I/10/Test.all_epitopes.tsv
+-rw-r--r--  2.0 unx   338645 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/example_data/MHC_Class_I/10/Test.all_epitopes.final.tsv
+-rw-r--r--  2.0 unx   234256 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/example_data/MHC_Class_I/10/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx     3890 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_I/10/Test.fa
 -rw-r--r--  2.0 unx    34243 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_I/10/Test.fa.tsv
--rw-r--r--  2.0 unx   318660 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/example_data/MHC_Class_I/8/Test.all_epitopes.final.tsv
--rw-r--r--  2.0 unx   218583 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/example_data/MHC_Class_I/8/Test.all_epitopes.tsv
+-rw-r--r--  2.0 unx   318660 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/example_data/MHC_Class_I/8/Test.all_epitopes.final.tsv
+-rw-r--r--  2.0 unx   218583 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/example_data/MHC_Class_I/8/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx     3784 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_I/8/Test.fa
 -rw-r--r--  2.0 unx    34243 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_I/8/Test.fa.tsv
--rw-r--r--  2.0 unx   333942 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/example_data/MHC_Class_I/9/Test.all_epitopes.final.tsv
--rw-r--r--  2.0 unx   231709 b- defN 23-Apr-05 14:24 pvactools/tools/pvacfuse/example_data/MHC_Class_I/9/Test.all_epitopes.tsv
+-rw-r--r--  2.0 unx   333942 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/example_data/MHC_Class_I/9/Test.all_epitopes.final.tsv
+-rw-r--r--  2.0 unx   231709 b- defN 23-Jun-06 18:07 pvactools/tools/pvacfuse/example_data/MHC_Class_I/9/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx     3837 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_I/9/Test.fa
 -rw-r--r--  2.0 unx    34243 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_I/9/Test.fa.tsv
 -rw-r--r--  2.0 unx     5485 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_II/Test.all_epitopes.aggregated.tsv
 -rw-r--r--  2.0 unx   222518 b- defN 22-May-20 15:09 pvactools/tools/pvacfuse/example_data/MHC_Class_II/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx     4156 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/MHC_Class_II/Test.fasta
 -rw-r--r--  2.0 unx     6588 b- defN 22-May-20 15:09 pvactools/tools/pvacfuse/example_data/MHC_Class_II/Test.filtered.tsv
 -rw-r--r--  2.0 unx   180200 b- defN 22-May-20 15:09 pvactools/tools/pvacfuse/example_data/MHC_Class_II/15/Test.all_epitopes.final.tsv
@@ -644,29 +644,29 @@
 -rw-r--r--  2.0 unx     1149 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/example_data/agfusion/ZNF660-44586213_CSNK2A1-508660/ZNF660-CSNK2A1.exons.txt
 -rw-r--r--  2.0 unx      230 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/example_data/agfusion/ZNF660-44586213_CSNK2A1-508660/ZNF660-CSNK2A1.fusion_transcripts.txt
 -rw-r--r--  2.0 unx      111 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/example_data/agfusion/ZNF660-44586213_CSNK2A1-508660/ZNF660-CSNK2A1.protein_domains.txt
 -rw-r--r--  2.0 unx     4505 b- defN 22-May-13 15:22 pvactools/tools/pvacfuse/example_data/agfusion/ZNF660-44586213_CSNK2A1-508660/ZNF660-CSNK2A1_cdna.fa
 -rw-r--r--  2.0 unx     5641 b- defN 22-May-17 15:14 pvactools/tools/pvacfuse/example_data/combined/Test.all_epitopes.aggregated.tsv
 -rw-r--r--  2.0 unx  1467372 b- defN 22-May-20 15:09 pvactools/tools/pvacfuse/example_data/combined/Test.all_epitopes.tsv
 -rw-r--r--  2.0 unx    10015 b- defN 22-May-20 15:09 pvactools/tools/pvacfuse/example_data/combined/Test.filtered.tsv
--rw-r--r--  2.0 unx      403 b- defN 23-Apr-05 14:24 pvactools/tools/pvacseq/__init__.py
+-rw-r--r--  2.0 unx      403 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/__init__.py
 -rw-r--r--  2.0 unx      376 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/allele_specific_cutoffs.py
 -rw-r--r--  2.0 unx      502 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/binding_filter.py
--rw-r--r--  2.0 unx      627 b- defN 23-Apr-05 14:24 pvactools/tools/pvacseq/calculate_reference_proteome_similarity.py
--rw-r--r--  2.0 unx     3646 b- defN 23-Apr-05 14:24 pvactools/tools/pvacseq/coverage_filter.py
+-rw-r--r--  2.0 unx      627 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/calculate_reference_proteome_similarity.py
+-rw-r--r--  2.0 unx     3646 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/coverage_filter.py
 -rw-r--r--  2.0 unx      374 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/download_example_data.py
--rw-r--r--  2.0 unx     1726 b- defN 23-Apr-05 14:24 pvactools/tools/pvacseq/generate_aggregated_report.py
--rw-r--r--  2.0 unx     6987 b- defN 23-Apr-05 14:24 pvactools/tools/pvacseq/generate_protein_fasta.py
+-rw-r--r--  2.0 unx     1726 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/generate_aggregated_report.py
+-rw-r--r--  2.0 unx     6987 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/generate_protein_fasta.py
 -rw-r--r--  2.0 unx     1007 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/install_vep_plugin.py
--rw-r--r--  2.0 unx     4352 b- defN 23-Apr-05 14:24 pvactools/tools/pvacseq/main.py
+-rw-r--r--  2.0 unx     4352 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/main.py
 -rw-r--r--  2.0 unx      377 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/net_chop.py
 -rw-r--r--  2.0 unx      392 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/netmhc_stab.py
--rw-r--r--  2.0 unx     8645 b- defN 23-Apr-05 14:24 pvactools/tools/pvacseq/run.py
--rw-r--r--  2.0 unx      382 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/top_score_filter.py
--rw-r--r--  2.0 unx     1580 b- defN 23-Apr-05 14:24 pvactools/tools/pvacseq/transcript_support_level_filter.py
+-rw-r--r--  2.0 unx     8645 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/run.py
+-rw-r--r--  2.0 unx      382 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/top_score_filter.py
+-rw-r--r--  2.0 unx     1580 b- defN 23-Jun-06 18:07 pvactools/tools/pvacseq/transcript_support_level_filter.py
 -rw-r--r--  2.0 unx      357 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/valid_alleles.py
 -rw-r--r--  2.0 unx     4689 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/VEP_plugins/Frameshift.pm
 -rw-r--r--  2.0 unx     1513 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/VEP_plugins/Wildtype.pm
 -rw-r--r--  2.0 unx   371831 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/example_data/input.vcf
 -rw-r--r--  2.0 unx    63283 b- defN 22-May-17 15:14 pvactools/tools/pvacseq/example_data/MHC_Class_I/Test.all_epitopes.aggregated.metrics.json
 -rw-r--r--  2.0 unx     3209 b- defN 22-May-17 15:14 pvactools/tools/pvacseq/example_data/MHC_Class_I/Test.all_epitopes.aggregated.tsv
 -rw-r--r--  2.0 unx   724354 b- defN 22-May-20 15:09 pvactools/tools/pvacseq/example_data/MHC_Class_I/Test.all_epitopes.tsv
@@ -722,15 +722,15 @@
 -rw-r--r--  2.0 unx   116941 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/iedb_alleles/class_ii/NetMHCIIpan.tsv
 -rw-r--r--  2.0 unx      407 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/iedb_alleles/class_ii/nn_align.tsv
 -rw-r--r--  2.0 unx      425 b- defN 22-May-13 15:22 pvactools/tools/pvacseq/iedb_alleles/class_ii/smm_align.tsv
 -rw-r--r--  2.0 unx      140 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/__init__.py
 -rw-r--r--  2.0 unx      379 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/allele_specific_cutoffs.py
 -rw-r--r--  2.0 unx      379 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/download_example_data.py
 -rw-r--r--  2.0 unx     1709 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/main.py
--rw-r--r--  2.0 unx    24079 b- defN 23-Apr-05 14:26 pvactools/tools/pvacvector/run.py
+-rw-r--r--  2.0 unx    24079 b- defN 23-Jun-06 18:07 pvactools/tools/pvacvector/run.py
 -rw-r--r--  2.0 unx      360 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/valid_alleles.py
 -rw-r--r--  2.0 unx     1304 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/visualize.py
 -rw-r--r--  2.0 unx      366 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/input.fa
 -rw-r--r--  2.0 unx     2581 b- defN 22-May-20 15:09 pvactools/tools/pvacvector/example_data/input.tsv
 -rw-r--r--  2.0 unx   207820 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/input.vcf
 -rw-r--r--  2.0 unx      567 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/results_from_fasta/Test_results.fa
 -rw-r--r--  2.0 unx     2985 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/results_from_fasta/0/AAY/junction_scores.best_spacers.tsv
@@ -753,35 +753,35 @@
 -rw-r--r--  2.0 unx     2798 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/results_from_tsv/0/GPGPG/junction_scores.best_spacers.tsv
 -rw-r--r--  2.0 unx    13834 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/results_from_tsv/0/GPGPG/junction_scores.tsv
 -rw-r--r--  2.0 unx     2793 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/results_from_tsv/0/HHHH/junction_scores.best_spacers.tsv
 -rw-r--r--  2.0 unx     8296 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/results_from_tsv/0/HHHH/junction_scores.tsv
 -rw-r--r--  2.0 unx     2834 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/results_from_tsv/0/None/junction_scores.best_spacers.tsv
 -rw-r--r--  2.0 unx     2834 b- defN 22-May-13 15:22 pvactools/tools/pvacvector/example_data/results_from_tsv/0/None/junction_scores.tsv
 -rw-r--r--  2.0 unx       42 b- defN 22-May-13 15:22 pvactools/tools/pvacview/__init__.py
--rw-r--r--  2.0 unx     9321 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/anchor_and_helper_functions.R
--rw-r--r--  2.0 unx       68 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/app.R
+-rw-r--r--  2.0 unx     9321 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/anchor_and_helper_functions.R
+-rw-r--r--  2.0 unx       68 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/app.R
 -rw-r--r--  2.0 unx      743 b- defN 22-May-13 15:22 pvactools/tools/pvacview/main.py
--rw-r--r--  2.0 unx      847 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/run.py
--rw-r--r--  2.0 unx    43105 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/server.R
--rw-r--r--  2.0 unx     3664 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/styling.R
--rw-r--r--  2.0 unx    17559 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/ui.R
+-rw-r--r--  2.0 unx      847 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/run.py
+-rw-r--r--  2.0 unx    43105 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/server.R
+-rw-r--r--  2.0 unx     3664 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/styling.R
+-rw-r--r--  2.0 unx    17559 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/ui.R
 -rw-r--r--  2.0 unx     6148 b- defN 22-May-13 15:22 pvactools/tools/pvacview/data/.DS_Store
--rw-r--r--  2.0 unx  1004292 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/data/H_NJ-HCC1395-HCC1395.all_epitopes.aggregated.metrics.json
--rw-r--r--  2.0 unx    46290 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/data/H_NJ-HCC1395-HCC1395.all_epitopes.aggregated.tsv
+-rw-r--r--  2.0 unx  1004292 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/data/H_NJ-HCC1395-HCC1395.all_epitopes.aggregated.metrics.json
+-rw-r--r--  2.0 unx    46290 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/data/H_NJ-HCC1395-HCC1395.all_epitopes.aggregated.tsv
 -rw-r--r--  2.0 unx    38268 b- defN 22-May-13 15:22 pvactools/tools/pvacview/data/Normalized_anchor_predictions_10_mer.tsv
 -rw-r--r--  2.0 unx    37432 b- defN 22-May-13 15:22 pvactools/tools/pvacview/data/Normalized_anchor_predictions_11_mer.tsv
 -rw-r--r--  2.0 unx    27396 b- defN 22-May-13 15:22 pvactools/tools/pvacview/data/Normalized_anchor_predictions_8_mer.tsv
 -rw-r--r--  2.0 unx    38170 b- defN 22-May-13 15:22 pvactools/tools/pvacview/data/Normalized_anchor_predictions_9_mer.tsv
--rw-r--r--  2.0 unx     4666 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/data/test_data_class_I.tsv
--rw-r--r--  2.0 unx     5740 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/data/test_data_class_II.tsv
--rw-r--r--  2.0 unx  7049710 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/data/test_data_class_II_metrics.json
--rw-r--r--  2.0 unx   460052 b- defN 23-Apr-05 14:24 pvactools/tools/pvacview/data/test_data_class_I_metrics.json
+-rw-r--r--  2.0 unx     4666 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/data/test_data_class_I.tsv
+-rw-r--r--  2.0 unx     5740 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/data/test_data_class_II.tsv
+-rw-r--r--  2.0 unx  7049710 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/data/test_data_class_II_metrics.json
+-rw-r--r--  2.0 unx   460052 b- defN 23-Jun-06 18:07 pvactools/tools/pvacview/data/test_data_class_I_metrics.json
 -rw-r--r--  2.0 unx   956623 b- defN 22-May-13 15:22 pvactools/tools/pvacview/www/anchor.jpg
 -rw-r--r--  2.0 unx     3667 b- defN 22-May-13 15:22 pvactools/tools/pvacview/www/pVACview_logo.png
 -rw-r--r--  2.0 unx     3356 b- defN 22-May-13 15:22 pvactools/tools/pvacview/www/pVACview_logo_mini.png
--rw-r--r--  2.0 unx     1692 b- defN 23-Apr-05 14:30 pvactools-3.1.2.dist-info/LICENSE
--rw-r--r--  2.0 unx     1303 b- defN 23-Apr-05 14:30 pvactools-3.1.2.dist-info/METADATA
--rw-r--r--  2.0 unx       92 b- defN 23-Apr-05 14:30 pvactools-3.1.2.dist-info/WHEEL
--rw-r--r--  2.0 unx      289 b- defN 23-Apr-05 14:30 pvactools-3.1.2.dist-info/entry_points.txt
--rw-r--r--  2.0 unx       10 b- defN 23-Apr-05 14:30 pvactools-3.1.2.dist-info/top_level.txt
-?rw-rw-r--  2.0 unx   117476 b- defN 23-Apr-05 14:30 pvactools-3.1.2.dist-info/RECORD
-785 files, 35260770 bytes uncompressed, 9774768 bytes compressed:  72.3%
+-rw-r--r--  2.0 unx     1692 b- defN 23-Jun-06 18:12 pvactools-3.1.3.dist-info/LICENSE
+-rw-r--r--  2.0 unx     1276 b- defN 23-Jun-06 18:12 pvactools-3.1.3.dist-info/METADATA
+-rw-r--r--  2.0 unx       92 b- defN 23-Jun-06 18:12 pvactools-3.1.3.dist-info/WHEEL
+-rw-r--r--  2.0 unx      288 b- defN 23-Jun-06 18:12 pvactools-3.1.3.dist-info/entry_points.txt
+-rw-r--r--  2.0 unx       10 b- defN 23-Jun-06 18:12 pvactools-3.1.3.dist-info/top_level.txt
+-rw-rw-r--  2.0 unx   117476 b- defN 23-Jun-06 18:12 pvactools-3.1.3.dist-info/RECORD
+785 files, 35260895 bytes uncompressed, 9774779 bytes compressed:  72.3%
```

## zipnote {}

```diff
@@ -2331,26 +2331,26 @@
 
 Filename: pvactools/tools/pvacview/www/pVACview_logo.png
 Comment: 
 
 Filename: pvactools/tools/pvacview/www/pVACview_logo_mini.png
 Comment: 
 
-Filename: pvactools-3.1.2.dist-info/LICENSE
+Filename: pvactools-3.1.3.dist-info/LICENSE
 Comment: 
 
-Filename: pvactools-3.1.2.dist-info/METADATA
+Filename: pvactools-3.1.3.dist-info/METADATA
 Comment: 
 
-Filename: pvactools-3.1.2.dist-info/WHEEL
+Filename: pvactools-3.1.3.dist-info/WHEEL
 Comment: 
 
-Filename: pvactools-3.1.2.dist-info/entry_points.txt
+Filename: pvactools-3.1.3.dist-info/entry_points.txt
 Comment: 
 
-Filename: pvactools-3.1.2.dist-info/top_level.txt
+Filename: pvactools-3.1.3.dist-info/top_level.txt
 Comment: 
 
-Filename: pvactools-3.1.2.dist-info/RECORD
+Filename: pvactools-3.1.3.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## pvactools/lib/calculate_reference_proteome_similarity.py

```diff
@@ -197,17 +197,22 @@
         flanking_sequence_length = self.match_length - 1
         start = subpeptide_position - 1 - flanking_sequence_length
         if start < 0:
             start = 0
         #This catches cases where the start position would cause too many leading wildtype amino acids, which would result
         #in false-positive reference matches
         if len(full_peptide) > len(wt_peptide):
-            diff_position = [i for i in range(len(wt_peptide)) if wt_peptide[i] != full_peptide[i]][0]
+            diffs = [i for i in range(len(wt_peptide)) if wt_peptide[i] != full_peptide[i]]
+            if diffs == []:
+                diffs = [len(wt_peptide)]
         else:
-            diff_position = [i for i in range(len(full_peptide)) if wt_peptide[i] != full_peptide[i]][0]
+            diffs = [i for i in range(len(full_peptide)) if wt_peptide[i] != full_peptide[i]]
+            if diffs == []:
+                diffs = [len(full_peptide)]
+        diff_position = diffs[0]
         min_start = diff_position - self.match_length + 1 
         if min_start > start:
             start = min_start
         end = start + flanking_sequence_length + len(epitope) + flanking_sequence_length
         return full_peptide[start:end]
```

## Comparing `pvactools-3.1.2.dist-info/LICENSE` & `pvactools-3.1.3.dist-info/LICENSE`

 * *Files identical despite different names*

## Comparing `pvactools-3.1.2.dist-info/METADATA` & `pvactools-3.1.3.dist-info/METADATA`

 * *Files 3% similar despite different names*

```diff
@@ -1,17 +1,16 @@
 Metadata-Version: 2.1
 Name: pvactools
-Version: 3.1.2
+Version: 3.1.3
 Summary: A cancer immunotherapy tools suite
 Home-page: https://github.com/griffithlab/pVACtools
 Author: Jasreet Hundal, Susanna Kiwala, Joshua McMichael, Yang-Yang Feng, Christopher A. Miller, Aaron Graubert, Amber Wollam, Connor Liu, Jonas Neichin, Megan Neveau, Jason Walker, Elaine R. Mardis, Obi L. Griffith, Malachi Griffith
 Author-email: help@pvactools.org
 License: BSD-3-Clause-Clear
 Keywords: antigens neoantigens cancer sequencing variant variants fusion fusions
-Platform: UNKNOWN
 Classifier: Development Status :: 4 - Beta
 Classifier: Intended Audience :: Science/Research
 Classifier: Topic :: Scientific/Engineering :: Bio-Informatics
 Classifier: Programming Language :: Python :: 3.6
 License-File: LICENSE
 Requires-Dist: vcfpy
 Requires-Dist: requests
@@ -29,9 +28,7 @@
 Requires-Dist: keras (==2.4.3)
 Requires-Dist: tensorflow (==2.2.2)
 Requires-Dist: protobuf (==3.19.4)
 Requires-Dist: mhcnuggets (==2.3.3)
 Requires-Dist: mhcflurry (==2.0.1)
 Requires-Dist: testfixtures
 
-UNKNOWN
-
```

## Comparing `pvactools-3.1.2.dist-info/RECORD` & `pvactools-3.1.3.dist-info/RECORD`

 * *Files 1% similar despite different names*

```diff
@@ -1,14 +1,14 @@
 pvactools/lib/__init__.py,sha256=t13fCY-BQeyISdhFkAt5WBretAfC2BEETfSAdCjd3oM,445
 pvactools/lib/aggregate_all_epitopes.py,sha256=Ya0RlE0-kRdaIMQC68mjjRkvgYieq7nQh2HwTsIA-HY,25560
 pvactools/lib/allele_specific_binding_filter.py,sha256=Vsxv5fsH7spBp95T5RRp_x2vRQ5l8JaWJeA_XuRcnE0,3099
 pvactools/lib/allele_specific_cutoffs.py,sha256=EsIdYERZHBkIdXN9Uasj3L7UXqH6yaMkbFZSOOkbkJ8,792
 pvactools/lib/binding_filter.py,sha256=59zlEseZkyWRm9r_FBZd8SC9iHFYeXlgLEYqPkVqCRE,5537
 pvactools/lib/calculate_manufacturability.py,sha256=3hyKpsfHo9AbG6uYBFyH4GCepPniO3cQEsDdP_Aqei8,3113
-pvactools/lib/calculate_reference_proteome_similarity.py,sha256=aycrEEBlrN845LtpfJBGp_oeQh8xfYSQusEEehEoVAs,17927
+pvactools/lib/calculate_reference_proteome_similarity.py,sha256=Y5Aocmjo7ni43D6BrBIZfj9wWJVhtR3cCjOxJs9Mt_Q,18080
 pvactools/lib/call_iedb.py,sha256=fWV7HaooloxPtcquJ5U4tLH9GkYsTYAIKYl8hLOV7W4,2326
 pvactools/lib/call_mhcnuggets.py,sha256=Ik0EUOQ7nRpmdmaPDqgVqVGXF3CO0Sa4BpyILFZvc_8,2981
 pvactools/lib/combine_parsed_outputs.py,sha256=geiEH-w4FXeiui4fuKxo110_klsx_woSyX4dx5ccA8E,2256
 pvactools/lib/csq_parser.py,sha256=YjhEZSbI_H0ad9_DY4-sijaE6LsTczxk-C3Vtc9mUQU,1523
 pvactools/lib/download_example_data.py,sha256=TSMycS_UwuGI-Emj0xiIU9xzagddoV2eGl0pBxzlQxE,955
 pvactools/lib/fasta_generator.py,sha256=PEHOkWEzVzcs1fVe-FIfDQN8AD3__r2Ta4SOKa0ew90,22842
 pvactools/lib/filter.py,sha256=dAHuG1rb1ukw0KBjJXeLCCkT0CVgmj8vOjMBR8xRMmU,1337
@@ -773,13 +773,13 @@
 pvactools/tools/pvacview/data/test_data_class_I.tsv,sha256=GS4VOLvVfO4qWRXVr8vdTyHvoiE0SzpjAJ0ZeNxlP5M,4666
 pvactools/tools/pvacview/data/test_data_class_II.tsv,sha256=YYMJEcKYoUjkrxOhpS-uCMwxZVSqyjCmkihptG3p4Ts,5740
 pvactools/tools/pvacview/data/test_data_class_II_metrics.json,sha256=lS6svVM0hHlIrlTOgTFqGzws2J5kCYXcbTXVof-xFGg,7049710
 pvactools/tools/pvacview/data/test_data_class_I_metrics.json,sha256=3UwXnXTxkA9-KM5vNRY3UU8cfoAyC3BRLFWpEw-j5vo,460052
 pvactools/tools/pvacview/www/anchor.jpg,sha256=LBvsBNu-kzr85ibTS7A8fMfv4Q2l12ePTfMkXuhY-9E,956623
 pvactools/tools/pvacview/www/pVACview_logo.png,sha256=6U5P4qtQRkjjSootB0FYoQqUIEwOXoKpjqehY21OapI,3667
 pvactools/tools/pvacview/www/pVACview_logo_mini.png,sha256=yYdm5DGHYlggJLGdldejLSM6YDSroMDwdwmbJFEQe8Q,3356
-pvactools-3.1.2.dist-info/LICENSE,sha256=SnigGJMm_MplPC0F1hk4329aDmMhYxZIr9Gwga11jMw,1692
-pvactools-3.1.2.dist-info/METADATA,sha256=DBIJ74_g6EUHjHmgTLyEM-xbLGMjy3PCzkyCzEin1lc,1303
-pvactools-3.1.2.dist-info/WHEEL,sha256=ewwEueio1C2XeHTvT17n8dZUJgOvyCWCt0WVNLClP9o,92
-pvactools-3.1.2.dist-info/entry_points.txt,sha256=yaqZpdRoub3H-avLYQ-EdglKU4-hUys5j8davJQmoGw,289
-pvactools-3.1.2.dist-info/top_level.txt,sha256=9fM-uBaGPwDl65QCw1GEDfAWP9iYqyPUo-3r4XA06VY,10
-pvactools-3.1.2.dist-info/RECORD,,
+pvactools-3.1.3.dist-info/LICENSE,sha256=SnigGJMm_MplPC0F1hk4329aDmMhYxZIr9Gwga11jMw,1692
+pvactools-3.1.3.dist-info/METADATA,sha256=XNKTSfn4GjcZMx_md5KFTuxx4o3GQHPNwpmXVZf7w0w,1276
+pvactools-3.1.3.dist-info/WHEEL,sha256=pkctZYzUS4AYVn6dJ-7367OJZivF2e8RA9b_ZBjif18,92
+pvactools-3.1.3.dist-info/entry_points.txt,sha256=sunI84H_UvLa5PHpjJcSboko3L0DzX0ib3CUwcjrUYo,288
+pvactools-3.1.3.dist-info/top_level.txt,sha256=9fM-uBaGPwDl65QCw1GEDfAWP9iYqyPUo-3r4XA06VY,10
+pvactools-3.1.3.dist-info/RECORD,,
```

